### ISSUE TITLE
udp_max_queries value is not reflecting

### ISSUE THREAD

**Issue by amartya321** [envoyproxy/envoy#37734]
*Title*: *udp_max_queries value is not changing*

*Description*:
I am using Envoy as a Dynamic Forward Proxy and Envoy is running in a pod within a AKS Cluster. 
In the Envoy Config I have set the udp_max_queries as 15, as shown below and same in the dynamic forward Proxy cluster.
![image](https://github.com/user-attachments/assets/54b79d61-ea2c-4470-aab7-951bf9772362)

I have two Envoy pod running and traffic is sent to these Pod from a load balancer K8s service. 

I have installed tcpdump in one of the Envoy pod to get pcap and was expecting for DNS query the UDP source port will change after every 15 request. 

![image](https://github.com/user-attachments/assets/95460a16-7992-4abf-92d2-2272e958da4c)

But here we can see udp source port is changing after 5 request. 
I have tried with udp_max_queries  set to 10, but got the same behavior. 

I have used latest envoy v1.32.2.

*Repro steps*:
Use Envoy as dynamic forward proxy and use c-ares as dns resolve and set udp_max_queries  to 15. 
Take pcap from one of the Envoy pods and see the behaviour.

*Config*:
This is part of static resource

    static_resources:
      listeners:
      - name: listener_0
        address:
          socket_address:
            address: 0.0.0.0
            protocol: TCP
            port_value: 3128
        filter_chains:
        - filters:
          - name: envoy.filters.network.http_connection_manager
            typed_config:
              '@type': type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
              stat_prefix: egress_http
              http2_protocol_options:
                allow_connect: true
              http_protocol_options:
                accept_http_10: true
                default_host_for_http_10: nfa-envoy
              upgrade_configs:
                - upgrade_type: CONNECT
              rds:
                route_config_name: local_route
                config_source:
                  resource_api_version: V3
                  api_config_source:
                    api_type: GRPC
                    transport_api_version: V3
                    grpc_services:
                      envoy_grpc:
                        cluster_name: xds_cluster
              http_filters:
              - name: envoy.filters.http.rbac
                typed_config:
                  '@type': type.googleapis.com/envoy.extensions.filters.http.rbac.v3.RBAC
              - name: envoy.filters.http.dynamic_forward_proxy
                typed_config:
                  '@type': type.googleapis.com/envoy.extensions.filters.http.dynamic_forward_proxy.v3.FilterConfig
                  dns_cache_config:
                    name: dynamic_forward_proxy_cache_config
                    dns_lookup_family: ALL
                    typed_dns_resolver_config:
                      name: envoy.network.dns_resolver.cares
                      typed_config:
                        "@type": type.googleapis.com/envoy.extensions.network.dns_resolver.cares.v3.CaresDnsResolverConfig
                        udp_max_queries: 
                          value: 15
              - name: envoy.filters.http.router
                typed_config:
                  '@type': type.googleapis.com/envoy.extensions.filters.http.router.v3.Router
                  
                  
And this is part of cluster config 

      clusters:
      - name: dynamic_forward_proxy_cluster
        # connect_timeout - this defines how much envoy will wait for connection to establish
        connect_timeout: 3s
        lb_policy: CLUSTER_PROVIDED
        circuit_breakers:
          thresholds:
            max_connections: {{ .Values.circuitBreaker.maxConnections }}
            max_pending_requests: {{ .Values.circuitBreaker.maxPendingRequests }}
            max_requests: {{ .Values.circuitBreaker.maxRequests }}
        cluster_type:
          name: envoy.clusters.dynamic_forward_proxy
          typed_config:
            '@type': type.googleapis.com/envoy.extensions.clusters.dynamic_forward_proxy.v3.ClusterConfig
            dns_cache_config:
              name: dynamic_forward_proxy_cache_config
              dns_lookup_family: ALL
              typed_dns_resolver_config:
                name: envoy.network.dns_resolver.cares
                typed_config:
                  "@type": type.googleapis.com/envoy.extensions.network.dns_resolver.cares.v3.CaresDnsResolverConfig
                  udp_max_queries: 
                    value: 15
        upstream_connection_options:
          tcp_keepalive:
            keepalive_probes: 5
            keepalive_time: 600  # 10 min in seconds
            keepalive_interval: 60

Do let me know If you need more information on this. 

Thank You

**Comment by adisuissa** [comment]
cc @deveshkandpal1224 who introduced this feature

**Comment by Stevenjin8** [comment]
@amartya321 AFAIK there are two reason for cares to close a udp socket and open a new one:
1. it has performed `udp_max_queries` queries on that socket
2. there are no active dns queries and [`ARES_FLAG_STAYOPEN`](https://c-ares.org/docs/ares_init_options.html) is not set.

Its hard to tell without timestamps, but my guess is that the 10 dns queries overlap.
After the 10th query, there are no more active queries, so cares will close the udp socket 60197 even though there have only been 10 queries. When running the next query, cares will create a new udp socket with a system-assigned port. If you really want to see `udp_max_queries` in action, you can set it to `1` and see that consecutive A and AAAA queries will be from different source ports.

TLDR, `udp_max_queries` is a maximum, not a minimum. There are other reasons that cause cares to switch udp ports.
